README.md
# Discord-Invite-Tracker

<a href="https://discord.gg/nPwjaJk"><img src="https://discord.com/api/guilds/757966278936756345/embed.png"/></a>
<a href="https://github.com/GregTCLTK/Discord-Invite-Tracker/fork"><img src="https://img.shields.io/badge/PRs-welcome-brightgreen.svg?style=flat-square" /></a>
## Getting started
**1.** Install discord.py (rewrite): `python3 -m pip install -U git+https://github.com/Rapptz/discord.py`

**2.** [Download](https://github.com/GregTCLTK/Discord-Invite-Tracker/archive/master.zip) & [Unzip](https://www.7-zip.org) the Project

**3.** Open `config.json` and fill in Discord Bot Token, Server ID and Logs Channel ID.

       --> Open the [Developer Portal](https://discord.com/developers/applications/me/create) & follow the [Tutorial](https://i.imgur.com/vR4C6Mh.gif)\
       --> Additionaly activate the `SERVER MEMBERS INTENT` for the bot

**4.** Usage
#### Bot
Start the Bot:

`$python3 bot.py`

#### Cog
Add Cog to bot
```py
bot.load_extension("invite_tracker")
#in your main bot file
```




### Any Questions?
-> Join our [Discord](https://discord.gg/nPwjaJk) and ask in #general and link this repository


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).
This project mostly adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html);

# v1.1.2

### Internal
bot.py cleaned up abit

# v1.1.1

## Added
### Internal
Invite Tracker Cog introduced (15th April 2021)


bot.py
# Copyright: GregTCLTK 2018-2021.
# Contact Developer on https://discord.gg/nPwjaJk (Skidder#8515 | 401817301919465482)

import discord
import asyncio
import json

intents = discord.Intents.default()
intents.members = True
client = discord.Client(intents=intents)

cfg = open("config.json", "r")
tmpconfig = cfg.read()
cfg.close()
config = json.loads(tmpconfig)

token = config["token"]
guild_id = config["server-id"]
logs_channel = config["logs-channel-id"]

invites = {}
last = ""


async def fetch():
    global last
    global invites
    await client.wait_until_ready()
    gld = client.get_guild(int(guild_id))
    logs = client.get_channel(int(logs_channel))
    while True:
        invs = await gld.invites()
        tmp = []
        for i in invs:
            for s in invites:
                if s[0] == i.code:
                    if int(i.uses) > s[1]:
                        usr = gld.get_member(int(last))
                        eme = discord.Embed(description="Just joined the server", color=0x03d692, title=" ")
                        eme.set_author(name=usr.name + "#" + usr.discriminator, icon_url=usr.avatar_url)
                        eme.set_footer(text="ID: " + str(usr.id))
                        eme.timestamp = usr.joined_at
                        eme.add_field(name="Used invite",
                                      value="Inviter: " + i.inviter.mention + " (`" + i.inviter.name + "#" + i.inviter.discriminator + "` | `" + str(i.inviter.id) + "`)\nCode: `" + i.code + "`\nUses: `" + str(
                                          i.uses) + "`", inline=False)
                        await logs.send(embed=eme)
            tmp.append(tuple((i.code, i.uses)))
        invites = tmp
        await asyncio.sleep(4)


@client.event
async def on_ready():
    print("ready!")
    await client.change_presence(activity=discord.Activity(name="joins", type=2))


@client.event
async def on_member_join(meme):
    global last
    last = str(meme.id)


client.loop.create_task(fetch())
client.run(token)


jointracker.py
import discord
import asyncio
import datetime
import json
import os

client = discord.Client()
cfg = open("config.json", "r")
tmpconfig = cfg.read()
cfg.close()
config = json.loads(tmpconfig)

token = config["token"]
guild_id = config["server-id"]
logs_channel = config["logs-channel-id"]


invites = {}
last = ""

async def fetch():
 global last
 global invites
 await client.wait_until_ready()
 gld = client.get_guild(int(guild_id))
 logs = client.get_channel(int(logs_channel))
 while True:
  invs = await gld.invites()
  tmp = []
  for i in invs:
   for s in invites:
    if s[0] == i.code:
     if int(i.uses) > s[1]:
      usr = gld.get_member(int(last))
      eme = discord.Embed(description = "Just joined the server", color = 0x03d692, title = " ")
      eme.set_author(name = usr.name + "#" + usr.discriminator, icon_url = usr.avatar_url)
      eme.set_footer(text = "ID: " + str(usr.id))
      eme.timestamp = usr.joined_at
      eme.add_field(name = "Used invite", value = "Inviter: " + i.inviter.mention + " (`" + i.inviter.name + "#" + i.inviter.discriminator + "`)\nCode: `" + i.code + "`\nUses: `" + str(i.uses) + "`", inline = False)
      await logs.send(embed = eme)
   tmp.append(tuple((i.code, i.uses)))
  invites = tmp
  await asyncio.sleep(4)


@client.event
async def on_ready():
 print("ready!")
 await client.change_presence(activity = discord.Activity(name = "joins", type = 2))


@client.event
async def on_member_join(meme):
 global last
 last = str(meme.id)




client.loop.create_task(fetch())
client.run(token)

invite_tracker.py
# Copyright: GregTCLTK 2018-2021.
# Contact Developer on https://discord.gg/nPwjaJk (Skidder#8515 | 401817301919465482)
# Cog by: Quill (quillfires)

import discord
import asyncio
import json
import time
import typing
import datetime
from discord.ext import commands
# from discord.ext.commands import has_permissions
from discord import Embed

cfg = open("config.json", "r")
tmpconfig = cfg.read()
cfg.close()
config = json.loads(tmpconfig)

class invite_tracker(commands.Cog):
    """
    Keep track of your invites
    """
    def __init__(self, bot):
        self.bot = bot
        self.logs_channel = config["logs-channel-id"]
        self.version = "1.0.0"

        self.invites = {}
        bot.loop.create_task(self.load())

    async def load(self):
        await self.bot.wait_until_ready()
        # load the invites
        for guild in self.bot.guilds:
            try:
                self.invites[guild.id] = await guild.invites()
            except:
                pass

    def find_invite_by_code(self, inv_list, code):
        for inv in inv_list:
            if inv.code == code:
                return inv

    @commands.Cog.listener()
    async def on_member_join(self, member):
        logs = self.bot.get_channel(int(self.logs_channel))
        eme = Embed(description="Just joined the server", color=0x03d692, title=" ")
        eme.set_author(name=str(member), icon_url=member.avatar_url)
        eme.set_footer(text="ID: " + str(member.id))
        eme.timestamp = member.joined_at
        try:
            invs_before = self.invites[member.guild.id]
            invs_after = await member.guild.invites()
            self.invites[member.guild.id] = invs_after
            for invite in invs_before:
                if invite.uses < self.find_invite_by_code(invs_after, invite.code).uses:
                    eme.add_field(name="Used invite",
                                  value=f"Inviter: {invite.inviter.mention} (`{invite.inviter}` | `{str(invite.inviter.id)}`)\nCode: `{invite.code}`\nUses: ` {str(invite.uses)} `", inline=False)
        except:
            pass
        await logs.send(embed=eme)

    @commands.Cog.listener()
    async def on_member_remove(self, member):
        logs = self.bot.get_channel(int(self.logs_channel))
        eme = Embed(description="Just left the server", color=0xff0000, title=" ")
        eme.set_author(name=str(member), icon_url=member.avatar_url)
        eme.set_footer(text="ID: " + str(member.id))
        eme.timestamp = member.joined_at
        try:
            invs_before = self.invites[member.guild.id]
            invs_after = await member.guild.invites()
            self.invites[member.guild.id] = invs_after
            for invite in invs_before:
                if invite.uses > self.find_invite_by_code(invs_after, invite.code).uses:
                    eme.add_field(name="Used invite",
                                  value=f"Inviter: {invite.inviter.mention} (`{invite.inviter}` | `{str(invite.inviter.id)}`)\nCode: `{invite.code}`\nUses: ` {str(invite.uses)} `", inline=False)
        except:
            pass
        await logs.send(embed=eme)

    @commands.Cog.listener()
    async def on_guild_join(self, guild):
        try:
            self.invites[guild.id] = await guild.invites()
        except:
            pass

    @commands.Cog.listener()
    async def on_guild_remove(self, guild):
        try:
            self.invites.pop(guild.id)
        except:
            pass


def setup(bot):
    bot.add_cog(invite_tracker(bot))
